### ISSUE TITLE
implementation: add provenance-preserving event features and leakage-safe splits

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#7386]
## Objective

Deliver the first event-table implementation for descriptive failure clustering under #7396: a versioned adapter, interpretable features, deterministic group splits, and leakage/missingness fixtures. Do not launch a clustering research campaign in this issue.

## Current scope and readiness — 2026-09-16

The maintainer requested implementation-ready work after completion of the [background research](https://github.com/ll7/robot_sf_ll7/issues/7386#issuecomment-5699737997). This issue now owns the **first implementation leaf**. #7396 remains the coordination and wider clustering/evaluation roadmap; it is not a reciprocal hard prerequisite.

The September 8 choice is retained: event-centred, interpretable features; existing manual taxonomy; training-only preprocessing; grouped validation; later k-means and agglomerative controls, with HDBSCAN optional. No additional literature review, model-selection decision, or human naming gate is required before this leaf can be implemented.

This is a manual issue-triage readiness decision under the current maintainer request. Automated live admission, code tests, and corpus evaluation were not run during this label update. The worker must use normal admission/claim checks and PR review; scientific conclusions remain separate.

```yaml
archetype: implementation
evidence_tier: idea
linked_policy:
  - docs/context/issue_1512_issue_archetypes.md
  - docs/context/artifact_evidence_vocabulary.md
```

## Inputs and existing owners

- `robot_sf/benchmark/failure_extractor.py` and `docs/dev/failure_extractor.md`: existing episode/failure extraction; do not change its thresholds or reinterpret its episode-level output as a timestamped event.
- `robot_sf/benchmark/mechanism_trace.py` and its versioned schema: trace-row provenance and typed diagnostic inputs where available.
- Existing taxonomy contracts reached through `docs/context/issue_2220_failure_mechanism_taxonomy.md`; preserve the operational taxonomy tracked by #191.
- #2754 and #7396 identify the wider failure-analysis and clustering owners.

The input is saved, source-addressable episode/event/trace data, not simulator access. Tiny committed fixtures are sufficient for implementation proof; a multi-scenario research corpus is not an entrance prerequisite.

## Scope and implementation plan

1. Add a small adapter in the existing benchmark package, proposed owner `robot_sf/benchmark/event_features.py`, with tests under `tests/benchmark/test_event_features.py`. Reuse existing event identities and records; do not create a second event detector or reclassify source failures.
2. Define rows keyed by parent run, scenario/template, seed, planner, event identity, actor set, and original event interval. Retain source artifact digest and row references. Where an episode record has no event timestamp, emit explicit event-window unavailability; never invent a window around the final outcome.
3. Define a versioned feature registry with name, units, source, transform, missingness, and planner-visible/post-hoc classification. The first supported features are event duration, recorded minimum clearance, recorded minimum TTC, collision/event count, recorded fallback count, and recorded control-saturation count. Source semantics control eligibility; an unavailable feature is null with a reason, not zero. An absent event flag counts as zero only when the source proves complete coverage.
4. Preserve all original taxonomy strings as metadata. Do not put run/seed/planner IDs, template names, source filenames, or hidden fixture truth into numeric clustering features. Do not expose a populated causal/root-cause field from descriptive grouping.
5. Implement group-aware split validation. Keep connected parent/snapshot/perturbation/near-duplicate lineages in one partition. Store explicit train/validation/test manifests; deterministic assignment uses the frozen group ID and seed, not row position. Conflicting declared partitions fail rather than silently moving test rows.
6. Add a fit/transform preprocessing boundary. Fit optional imputation and scaling only on training groups. All-missing training columns remain unsupported or are explicitly excluded in the fitted schema; test data cannot rescue them. Preserve missingness indicators and the exact fitted feature order.
7. Produce a compact JSON-safe table plus schema, split manifest, feature-availability report, and deterministic digest. Document the public Python adapter and a fixture-driven example. Keep the original records immutable.

## Outputs

The adapter, feature registry, table export, group-split validator, training-only preprocessing boundary, small fixtures, and indexed documentation. No fitted scientific clustering model or new learned representation is required for completion.

## Acceptance criteria

- [ ] Every output row resolves back to original source rows and an immutable parent lineage.
- [ ] Missing event timing and missing features remain explicit; non-finite values never appear as valid numeric JSON.
- [ ] Manual taxonomy values round-trip unchanged.
- [ ] Parent, derivative, and near-duplicate groups cannot cross splits.
- [ ] Changing row order does not alter identities, partitions, fitted feature order, or canonical outputs.
- [ ] Changing held-out values cannot change fitted preprocessing statistics.
- [ ] All-missing training features and unsupported units have explicit outcomes.
- [ ] Fixture truth, IDs, and future outcome labels cannot leak into an allegedly online or unlabeled feature set.
- [ ] Rare/noise/unknown cases remain retained and retrievable; causal fields stay unset.
- [ ] Existing failure-extractor and mechanism-trace tests remain compatible.

## Verification

```bash
uv run pytest tests/benchmark/test_event_features.py -q
uv run pytest tests/benchmark -k 'failure or mechanism_trace' -q
uv run ruff check robot_sf/benchmark/event_features.py tests/benchmark/test_event_features.py
uv run ruff format --check robot_sf/benchmark/event_features.py tests/benchmark/test_event_features.py
git diff --check
BASE_REF=origin/main scripts/dev/pr_ready_check.sh
```

The first file/test above is a proposed deliverable, not a claim that it already exists. Use the repository environment wrapper where required and update the commands if an existing equivalent owner is reused.

## Dependencies and non-goals

Hard implementation dependencies: none beyond the checked-in source contracts. Parent #7396 coordinates the later work; no need to close it before coding this adapter. Recheck live ownership and reuse any intervening equivalent component.

The later clustering comparison, grouped bootstrap stability, held-out assignment protocol, descriptive naming, optional HDBSCAN/time-series embeddings, and corpus-level conclusions remain under #7396. Its initial 100-resample protocol must resample parent groups and compare common identities; this leaf does not claim that evaluation has run. Controlled root-cause experiments, model training, simulator campaigns, benchmark changes, and manuscript claims are outside scope.

**Comment by ll7** [comment]
ll7/robot_sf_ll7#7386: select-option-a-for-preparation-only

Select event-centred windows with an interpretable standardized feature table for the preparation phase. Compare the existing rule taxonomy, k-means, agglomerative clustering, and HDBSCAN with explicit noise retention. Time-series, learned, and graph representations remain later extensions.

Reasoning: Option A is the most auditable way to establish provenance, leakage controls, cluster stability, rare-case behavior, and semantic-review boundaries. It supports descriptive grouping without smuggling in a causal claim. Confidence: 98%.

Automatic consequence: #7396 may prepare a multi-scenario/multi-planner sample table, feature provenance and scaling report, duplicate/leakage plan, synthetic known-mechanism recovery fixture, non-causal cluster example, and sample-density estimate. No cluster may be named as a cause.

Reapply `decision-required` when that packet exists to freeze the event window, exact schema, missing-data policy, normalization, stability thresholds, expert-review process, unknown semantics, and manual-taxonomy role.

Reopen the representation only if frozen recovery and stability evidence shows the interpretable table systematically loses a declared temporal mechanism; then test a time-series view as an additive comparator, not a silent replacement.

**Comment by ll7** [comment]
Phase-state reconciliation: the current preparation decision is resolved; Option A is selected. The complete representation and stability contract is deferred until #7396 posts the required evidence packet. The `decision-required` label is removed, and no author decision is currently pending.

**Comment by ll7** [comment]
ll7/robot_sf_ll7#7386: select-option-a-for-preparation-only

The event-centred, interpretable-feature preparation direction remains selected. #7396 is still parked and has not delivered the multi-scenario sample table, feature provenance/scaling report, leakage plan, synthetic recovery fixture, non-causal example, or sample-density evidence required for the final schema/stability decision.

Remove the premature decision routing and retain the parked research boundary. Reapply `decision-required` only when the complete packet exists. Cluster labels remain descriptive and may never populate causal or root-cause fields without separately identified evidence.

**Comment by ll7** [comment]
ll7/robot_sf_ll7#7386: approve-descriptive-event-clustering-v1

Delegated maintainer decision — 2026-09-08. The maintainer explicitly authorized the agent to choose and record the decision for this issue.

Select descriptive event-window clustering using the existing taxonomy and failure-extraction owners. Define a versioned feature table keyed by run, scenario, seed, planner and actor set, preserving units, source and missingness. Use grouped splits and fit preprocessing only on training data. Start with manual taxonomy, k-means and agglomerative controls; add HDBSCAN as an optional comparator if supported by the existing dependency profile. Retain noise and rare cases. Cluster names are descriptive and must not populate causal/root-cause fields.

Next action: Through #7396, implement the event-table adapter and fixture-based leakage/missingness tests first, then evaluate on the available multi-scenario, multi-planner corpus. Use 100 deterministic bootstrap resamples as an initial stability protocol. The agent may assign descriptive labels with trace examples and uncertainty. Keep outputs exploratory until measured stability and coverage support stronger use; no new human naming gate is required.

Reopen only if the chosen features erase an important temporal mechanism or measured clusters remain unstable or uninterpretable.

**Comment by ll7** [comment]
Reconcile (2026-09-14): removing stale `decision-required`. The decision was already recorded by delegated maintainer ruling on 2026-09-08 (see comments above); the `ruled` label stands. No scope or option changed by this edit.

**Comment by ll7** [comment]
State reconcile (2026-09-14): applying `state:blocked`. The 2026-09-08 delegated ruling stands (options already selected), but the parked-parent prerequisites noted 2026-08-31 are still missing and the parents (#7395/#7396/#7397, plus #7398 for #7388) are untouched since 2026-08-30. No execution-state label was present after the stale `decision-required` removal; `state:blocked` is the honest state until the parent outputs land.

**Comment by ll7** [comment]
## Background research — descriptive clustering without false causal labels

**Recommendation:** retain the manual-taxonomy, k-means and agglomerative controls; use HDBSCAN only as an optional comparator. The main contribution should be an interpretable event table and a credible stability/coverage report, not an automatically discovered “root-cause taxonomy.”

### Findings from the literature and current APIs

Clustering-stability research shows that stability is not, by itself, a reliable identification of the correct number or meaning of clusters; theoretical results depend on the objective and data distribution [1]. Scikit-learn's HDBSCAN distinguishes clustered points from noise and invalid/missing observations, and its `min_samples` convention differs from the separate `hdbscan` package [2]. Its documented public interface provides `fit`/`fit_predict`, not a standard inductive `predict` for new observations. An implementation must therefore name its package and held-out assignment policy rather than assume all HDBSCAN implementations are interchangeable.

### Proposed scientific contract

**Separate metadata, descriptions and labels.** Run/seed/planner/actor identities belong in provenance and grouping columns, not automatically in clustering features. Otherwise clusters may merely recover the planner identifier. Preserve the manual taxonomy as an external comparison, not an input feature that makes agreement tautological. Distinguish pre-event observable features from hindsight summaries.

**Preserve temporal information deliberately.** Minima and means alone can collapse “late braking after occlusion” and “persistent crowding” into similar vectors. Add a small, interpretable set of timing/order features only where the trace supports them: time to first fallback, duration below a clearance threshold, or ordering of braking and closest approach. Do not infer an unobserved mechanism from an aggregate value.

**Fit preprocessing on development data only.** Freeze feature definitions, missingness handling and scaling before held-out analysis. Compare feature ablations so one large-unit or high-variance family does not dominate distance. The existing 100 deterministic resamples are a starting stability protocol, not a guarantee of statistical power.

**Make stability comparisons meaningful.** Resample independent parent lineages rather than overlapping event windows. Compare partitions on the same uniquely identified observations, or their declared common subset, rather than comparing unrelated row positions. Adjusted Rand index is label-permutation invariant [3], but report sample coverage and cluster/noise transitions alongside it. Two all-noise outputs should not be presented as successful discovery just because their partitions agree.

**Do not force every event into a named group.** Retain rare, noise and missing-data cases. Descriptive names should cite representative traces and state what was observed. The delegated agent can assign those descriptions; this does not add a new human naming gate. Causal fields remain separate.

### Decision-ready evaluation output

For every method/configuration: number of usable parents/events, non-noise coverage, cluster-size distribution, resampling stability, manual-taxonomy association and representative/boundary/rare examples. Undefined metrics must be reported as unavailable, not zero. Compare the practical value of finding recurring interpretable patterns, not just a single optimized cluster score.

### Primary sources

[1] Ben-David, von Luxburg and Pál, *A Sober Look at Clustering Stability* — author-institution record: https://is.mpg.de/publications/4109

[2] Scikit-learn HDBSCAN API and implementation differences: https://scikit-learn.org/stable/modules/generated/sklearn.cluster.HDBSCAN.html

[3] Scikit-learn adjusted Rand index: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.adjusted_rand_score.html

Research checked 2026-09-16. Evaluation results remain to be produced through #7396; this synthesis does not establish that a stable or useful clustering already exists.
